docs/sc-memory/api/cpp/extended/agents/agents.md

```diff
@@ -0,0 +1,660 @@
+
+### **ScAgent**
+
+This class can be used for all types of platform-dependent agents. The example using this class is represented below.
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+// Inherit your agent class from `ScAgent` class and specify template argument 
+// as sc-event class. Here `ScEventGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>` 
+// is type of event to which the given agent reacts.
+class MyAgent : public ScAgent<ScEventGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>>
+{
+public:
+  // Here you should specify class of actions which the given agent performs. 
+  // This overriding is required.
+  ScAddr GetActionClass() const override;
+  // Here you should implement program of the given agent. 
+  // This overriding is required.
+  ScResult DoProgram(
+    ScEventGenerateIncomingArc<ScType::EdgeAccessConstPosPerm> const & event, 
+    ScAction & action) override;
+
+  // Other user-defined methods.
+};
+```
+
+You can't override `DoProgram` without sc-event argument. There can be override one of these methods. 
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+class MyAgent : public ScAgent<ScEventGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>>
+{
+public:
+  ScAddr GetActionClass() const override;
+  ScResult DoProgram(ScAction & action) override;
+
+  // Other user-defined methods.
+};
+```
+
+See [**C++ Events API**](events.md) and [**C++ Event subscriptions API**](event_subscriptions.md) to learn how to use and handle sc-events.
+
+!!! note
+    Define action class as keynode in agent class or keynodes class.
+
+!!! warning
+    You must override methods `GetActionClass` and `DoProgram`.
+
+You can specify any existing event types as a template argument to the `ScAgent` class. For example, you can create agent that will be triggered to sc-event of removing sc-element.
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+class MyAgent : public ScAgent<ScEventEraseElement>
+{
+public:
+  ScAddr GetActionClass() const override;
+  ScResult DoProgram(
+    ScEventEraseElement const & event, ScAction & action) override;
+
+  // Other user-defined methods.
+};
+```
+
+If you want to change specification of this agent in knowledge base, then write like this:
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+class MyAgent : public ScBaseAgent
+{
+public:
+  ScAddr GetActionClass() const override;
+  ScResult DoProgram(
+    ScElementaryEvent const & event, ScAction & action) override;
+
+  // Other user-defined methods.
+};
+```
+
+This implementation allows to provide any sc-event type to `DoProgram`.
+
+!!! note
+    `ScBaseAgent` is alias for `ScAgent<ScElementaryEvent>`.
+
+!!! note
+    If you provide specification of your agent in knowledge base, then you don't need to override `GetActionClass`.
```

Comment:
I can write an algorithm for loading a specification for a given agent. I am afraid that it should not be difficult for the user

Reply:
I'm talking about something a little different. It's necessary to clarify how to put the specification in the DB so that in the C++ implementation of the agent it would be clear that this is the specification of this agent.

Reply:
I will leave link to ScModule API here and write guide for writing agent with specification in knowledge base in separate PR

Reply:
Fixed